shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -81,6 +118,16 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
     private def containsMth(key: (Str, Str) \/ (Opt[Str], Str)): Bool = mthEnv.contains(key) || parent.exists(_.containsMth(key))
     def containsMth(parent: Opt[Str], nme: Str): Bool = containsMth(R(parent, nme))
     def nest: Ctx = copy(Some(this), MutMap.empty, MutMap.empty)
+    // * Enter a new quoted environment with empty `qenv` and `fvars`.
+    // * A whole quasiquote (i.e., code"...") contains no binding or free variables at the beginning.
+    // * For a quoted binding (e.g., code"x => ..."):
+    // ** 1. The context of the lambda body is still in the quotation so `inQuote = true`
+    // ** 2. `qenv` only contains skolems created by the current binding and `fvars` only contains free variables appearing in the body.
+    // ** So we also apply empty `qenv` and `fvars` at the beginning.
+    // ** e.g. `code"x => y => x + y"`. For `y => x + y`, fvars = {\ga_x, \ga_y}, qenv = {\ga_y}. 
+    // ** So for `code"x => ..."`, fvars = {\al}, qenv = {\ga_x}, where \ga_x \/ \ga_y <= \al \/ \ga_y
+    def enterQuote: Ctx = copy(Some(this), MutMap.empty, MutMap.empty, inQuote = true, qenv = MutMap.empty, fvars = MutSet.empty)
```

Comment:
The name (which I suggested based on a misunderstanding) is inappropriate. This is not JUST for entering quotes. You should use critical thinking and tell me so instead of obliging and using an inadequate name. A better name could be `enterQuotedScope`.